zerver/lib/actions.py

```diff
@@ -909,6 +909,11 @@ def update_users_in_full_members_system_group(
         bulk_add_members_to_user_group(full_members_system_group, new_full_members)
 
 
+def promote_new_full_members() -> None:  # nocoverage
+    for realm in Realm.objects.all():
+        update_users_in_full_members_system_group(realm)
```

Comment:
This should be `Realm.objects.filter(deactivated=False).exclude(waiting_period_threshold=None), right?  

I think we can write a great test for this function by mocking time.  But the setup is basically going to be:
(1) Hand-remove a user from the full members system group, and adding another who doesn't belong.
(2) Deactivate the realm, and verify calling this does nothing.
(3) Reactivate the realm, and verify that calling this function corrects this.

Reply:
Related question: What do we want to happen with deactivated user accounts?  

The current logic updates their "Full member" status, which may be reasonable, but deserves at least a comment?  Maybe worth a bit of an API design discussion on that detail.

Reply:
> This should be `Realm.objects.filter(deactivated=False).exclude(waiting_period_threshold=0), right?

Done this change and also added the test.

> Related question: What do we want to happen with deactivated user accounts?
> 
> The current logic updates their "Full member" status, which may be reasonable, but deserves at least a comment? Maybe worth a bit of an API design discussion on that detail.

The `date_joined` field is still the same after deactivating and then reactivating the user. So, I think we should update the Full member status as well so the user is in correct group after reactivation. Can add a comment once we finally come on a conclusion.

Reply:
Yeah, I think just calling the "update state" event after reactivation sounds correct.

Reply:
Just to confirm, what you mean by event here, do we want to update the user group of user only after reactivation?